src/node/miner.cpp

```diff
@@ -59,8 +59,11 @@ void RegenerateCommitments(CBlock& block, ChainstateManager& chainman)
 
 static BlockAssembler::Options ClampOptions(BlockAssembler::Options options)
 {
-    // Limit weight to between 4K and DEFAULT_BLOCK_MAX_WEIGHT for sanity:
-    options.nBlockMaxWeight = std::clamp<size_t>(options.nBlockMaxWeight, 4000, DEFAULT_BLOCK_MAX_WEIGHT);
+    Assume(options.coinbase_max_additional_weight <= DEFAULT_BLOCK_MAX_WEIGHT);
+    Assume(options.coinbase_output_max_additional_sigops <= MAX_BLOCK_SIGOPS_COST);
```

Comment:
In commit "Pass coinbase constraints to createNewBlock" (e67e4669a9e263d90f0e276c8e65e93c39170375)

Current code seems ok, but depending on how important it is for this not to generate an invalid block, it may make sense to use `assert` instead of `Assume`, since Assume does not do anything in release builds. Or this could also abort in debug builds but cap the values in release builds:

```c++
if (!Assume(options.coinbase_output_max_additional_sigops <= MAX_BLOCK_SIGOPS_COST)) {
    options.coinbase_output_max_additional_sigops = MAX_BLOCK_SIGOPS_COST
}
```

Reply:
Currently neither of these values can be configured. With Stratum v2 I intend to add runtime checks. If those checks are wrong then I assume the fuzzer would eventually find this `Assume`?

Reply:
re: https://github.com/bitcoin/bitcoin/pull/30356#discussion_r1679851292

So you are saying if the values are too big at this point, it means there is a bug elsewhere in the code. If that is the case, I think it would be more appropriate to use `assert` or `Assert` rather than `Assume` so the bug does not go undetected until it is too late and eventually results in an invalid block being produced. I could be wrong, but I don't see a rationale for preferring assume over assert here.

Reply:
> So you are saying if the values are too big at this point, it means there is a bug elsewhere in the code

Yes

Changed to Assert.

Reply:
This is great! Despite that these values are hard coded, I also think it is safer to prefer an assert over Assume in this case. In that way we still have a safeguard in place during the whole integration process of Stratum V2 and even if the future run time checks somehow fail to deliver their intended purpose at least we have some feedback in debug mode. On the other hand the assert wont hurt anyone exactly because the values are hard coded.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/30356#discussion_r1681361423

Just to clarify for other reviewers since the names of these macros are confusing: In bitcoin core, `assert` and `Assert` macros behave very differently from the `assert` macro in standard c++. In standard c++ `assert` statements are not executed at all in release builds, while in bitcoin core they are always executed and always cause the execution to abort if they fail. Other c++ libraries that don't to be confusing use the term [check](https://github.com/bitcoin/bitcoin/pull/16136#issuecomment-501779934) instead of assert for this behavior.

The `Assume` macro is more unusual, because it doesn't abort execution in release builds, but does also doesn't get skipped in release builds, so will always evaluate arguments passed to it. It can make sense to use Assume instead of assert when an unexpected condition indicates some kind of bug, but the bug is easy to recover from, so it is worth crashing debug builds but not release builds. Normally it make sense to check the `Assume` return value and do something like log a warning or fix invalid values, not just ignore the unexpected condition completely.